src/mml_arithmetic.tpp

```
@@ -102,4 +105,190 @@ void Arithmetic_mml<T>::elementwise_in_place(const shared_ptr<Tensor<T>> a, T (*
       indices[d] = 0;  // Carry over to the next dimension
     }
   }
+}
+
+template <typename T>
+std::shared_ptr<Tensor<T>> Arithmetic_mml<T>::reduce_max(
+    const std::shared_ptr<const Tensor<T>> input, int axis) const {
+
+    // Ensure the input tensor is valid
+    if (!input) {
+        throw std::invalid_argument("Input tensor cannot be null.");
+    }
+
+    auto shape = input->get_shape();
+
+    // Adjust negative axis values
+    if (axis < 0) {
+        axis += shape.size();
```

If the axis value is negative shouldnt we throw an error instead?

---

No since ONNX specify that "Describes the dimension Softmax will be performed on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input)." This reminded me that i need to make the default -1 as ONNX specify.